### ISSUE TITLE
[Bug]: search_iterator taking too much time to give response after some iteration

### ISSUE THREAD

**Issue by Basir-mahmood** [milvus-io/milvus#34816]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: 2.3.13
- Deployment mode(standalone or cluster): cluster
- MQ type(rocksmq, pulsar or kafka): pulsar 
- SDK version(e.g. pymilvus v2.0.0rc2): 2.4.4
- OS(Ubuntu or CentOS): Ubuntu
- Others: 
  -  Index-type : DiskANN
```

### Current Behavior

I am using search_iterator, to get records in pages, and I am using milvus diskANN index. However, after some iterations, sometimes 2 or sometime 5 iterations on different collection, the search_iterator give response taking too much time, even after several minutes, no response is searched, and it keeps on halt. 
I am performing search on the partitions. The batch-size is 500, and limit is 10000. Moreover, the search params are "metric_type": "COSINE", "search_list": 1000. 

For a collection with 2.6 million records, it give first iteration results in 0.03 milli seconds, and for the second iteration it gives results in 176 seconds ( almost 3 minutes). 
For another collection, with with almost 85 million records. It is giving top 5 iterations within milli seconds, but it get stuck afterwards, and I didnt get any response after even waiting for couple of minutes.

### Steps To Reproduce
```

test_collection = Collection(collection_name)

search_params =  {"metric_type": "COSINE",
                 "search_list": 1000,
                }

iterator = test_collection.search_iterator(
    batch_size=500,
    data=[embeddings],
	anns_field="embeddings",
	partition_names = ["parition_name"],
	param=search_params,
	limit= 10000,
    output_fields=["id"]
)
results = []
while True:
    result = iterator.next()
    if not result:
        iterator.close()
        break
```

**Comment by xiaofan-luan** [comment]
please help on reproduce this issue on latest 2.3.18

I would expect to see longer response time after some iteration, but shouldn't be longer than 1 seconds, unless you have very small range

**Comment by yanliang567** [comment]
/assign @Basir-mahmood 
please keep us posted if any updates on 2.3.18 or 2.4.6
/unassign

**Comment by Basir-mahmood** [comment]
Thanks for the quick reply. I want to ask that as the cluster is in the production how can I upgrade the cluster without affecting the user-experience. There is no release for helm for this version-2.3.18, so how can I define that ?
I am trying to follow this, link: https://milvus.io/docs/upgrade_milvus_cluster-helm.md

**Comment by yanliang567** [comment]
you can just update the milvus tag to the target milvus image tag in values.yaml

**Comment by xiaofan-luan** [comment]
@yanliang567 
did we have similar test in house?

I'm also interested on seeing the perf behaiour when we iterated to later stage on all the index types

**Comment by yanliang567** [comment]
we tested the batch size=1000/5000 to iterator all the entities in a 10_million dataset,  the latency increased for 2-4 times to the first iterator, but still less than 800ms. I attached a screenshot of batch_size=1000 for reference:
![image](https://github.com/user-attachments/assets/647104c4-8c9e-403f-b5e5-4678447fb0c5)

**Comment by xiaofan-luan** [comment]
> we tested the batch size=1000/5000 to iterator all the entities in a 10_million dataset, the latency increased for 2-4 times to the first iterator, but still less than 800ms. I attached a screenshot of batch_size=1000 for reference: ![image](https://private-user-images.githubusercontent.com/82361606/350809883-647104c4-8c9e-403f-b5e5-4678447fb0c5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2MTEwMTcsIm5iZiI6MTcyMTYxMDcxNywicGF0aCI6Ii84MjM2MTYwNi8zNTA4MDk4ODMtNjQ3MTA0YzQtOGM5ZS00MDNmLWI1ZTUtNDY3ODQ0N2ZiMGM1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIyVDAxMTE1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNjZjBiYmRhYjEwZWFlYzA0MmRlNmY2ZTIyMmE4MmZkZjdmMzVjZGVkOTRiN2NiOTFlMTBmNDhjNDkzZDE5NjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OXClXOznhlU9tW7jIGZ_BSmf4UtTk3_LFi4K4dEOyMQ)

is this diskann?

**Comment by yanliang567** [comment]
the test result about was on HNSW index. 
@Basir-mahmood does your querynodes running on nvme disk, as you are running with diskann index

**Comment by Basir-mahmood** [comment]
@yanliang567 yeah, the query nodes are running on the nvme disk.

**Comment by Basir-mahmood** [comment]
@yanliang567 
For changing the milvus-image-version, should I just change the following configs
```
image:
  all:
    repository: milvusdb/milvus
    tag: v2.3.18
    pullPolicy: IfNotPresent
```

`helm upgrade milvus-cluster-name  zilliztech/milvus -n milvus --reuse-values -f config.yml`

And can you please confirm that I would not have to change the milvus-helm version ?

**Comment by yanliang567** [comment]
> @yanliang567 For changing the milvus-image-version, should I just change the following configs
> 
> ```
> image:
>   all:
>     repository: milvusdb/milvus
>     tag: v2.3.18
>     pullPolicy: IfNotPresent
> ```
> 
> `helm upgrade milvus-cluster-name zilliztech/milvus -n milvus --reuse-values -f config.yml`
> 
> And can you please confirm that I would not have to change the milvus-helm version ?

yes, it is okay to update the tag

**Comment by Basir-mahmood** [comment]
I have updated the milvus cluster to milvus-2.3.20, and I am using pymilvus version 2.4.6. But still, the issue persist. The first iteration was completed in milliseconds and then the next one took around 19 minutes to run. There are approximately 105 million records in the collection.

**Comment by xiaofan-luan** [comment]
I don't think older version of diskann is good at iterations.

We have special implementation of iterator for HNSW.  but I'm not pretty sure if we have similar optimization on diskann.

@yanliang567 can we track on the iteration speed of diskann for 2.4.10?
@liliu-z could you confirm?

**Comment by xiaofan-luan** [comment]
> 我们测试了 batch size=1000/5000 来迭代 10_million 数据集中的所有实体，第一个迭代器的延迟增加了 2-4 倍，但仍然少于 800ms。我附上了 batch_size=1000 的屏幕截图以供参考： ![图像](https://private-user-images.githubusercontent.com/82361606/350809883-647104c4-8c9e-403f-b5e5-4678447fb0c5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU1ODIwNTMsIm5iZiI6MTcyNTU4MTc1MywicGF0aCI6Ii84MjM2MTYwNi8zNTA4MDk4ODMtNjQ3MTA0YzQtOGM5ZS00MDNmLWI1ZTUtNDY3ODQ0N2ZiMGM1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA2VDAwMTU1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMzMmNmYzk4OTFmZjcwNTFlMmY2ZTVmNmY3Y2JhYWM2YjQ1YmRlOWVlYWM4YTBjNDU2NGFhMzZiNThlY2JhYTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.t_Oqaapxki9Ucq8PlR7xogPt7AeME5Wij3PFJxrMbBA)

I think this is more of a result from HNSW

**Comment by xiaofan-luan** [comment]
by the way, what is the use case of your iterator? diskann is usually much slower on large topk. IVFSQ, IVFPQ could be waht you are looking for

**Comment by liliu-z** [comment]
The old version of DiskANN is not good at iteration/range search. The new version is in the progress and will be released very soon
/assign @alwayslove2013

**Comment by Basir-mahmood** [comment]
@xiaofan-luan We have vector database currently having vecors around half a billion, this is one of the collection that I mentioned. There are multiple collections in the database. We are also concerned about the accuracy. Thats why we dont want to go towards quantization based approaches. And due to large sized data, we also want to avoid memory based index. 

Thus, we want to stay with the DiskANN index, and we also want to extract all the records against a search_vector, which is greater than 16,000 ( max limit ). For the purpose, we were using search_iterator.
Is there any other way around to perform it. 
Or is there any way possible that we define our timebases search_space for searching only on the newly inserted data ( e.g., last 48 hours)

**Comment by xiaofan-luan** [comment]
1. you can probably use partitions and each partiton hold one datas data
2. we are working on performance improvement for diskann search, this could potentially solve your problem

**Comment by Basir-mahmood** [comment]
The maximum number of partitions is limited to 1024. If we would be storing on date, then this would limit us. I guess we have to go with the option of storing data weekly. 
Is there any other method that would be more appropriate in your opinion?

**Comment by xiaofan-luan** [comment]
you can actually change the limit to 4096, I thought it's a config can be changed

**Comment by xiaofan-luan** [comment]
Iterator is reimplemented for disk ann of Milvus 2.5 and will be more improvement soon

**Comment by NaruSaku** [comment]
@xiaofan-luan Is the optimized diskann search iterator in current 2.5.2 release now?

**Comment by alwayslove2013** [comment]
@NaruSaku Yes. Diskann search iterator has been optimized in [Knowhere PR#874](https://github.com/zilliztech/knowhere/pull/874) and has been released with Milvus 2.5.0.
